File: Leetcode 0106 Construct Binary Tree from Inorder and Postorder Traversal [medium] [tree]/Construct Binary Tree from Inorder and Postorder Traversal [divide-conquer].java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (inorder == null || postorder == null || inorder.length != postorder.length) {
            return null;
        }
        
        return buildTree(inorder, postorder, 0, 0, inorder.length);
    }
    
    private TreeNode buildTree(int[] inorder, int[] postorder, int inorderStart, int postorderStart, int size) {
        if (size == 0) {
            return null;
        }
        
        TreeNode newRoot = new TreeNode(postorder[postorderStart + size - 1]);
        
        int inorderIndex = inorderStart;
        while (inorder[inorderIndex] != postorder[postorderStart + size - 1]) {
            inorderIndex++;
        }
        
        int leftSize = inorderIndex - inorderStart;
        int rightSize = size - leftSize - 1;
        TreeNode leftChild = buildTree(inorder, postorder, inorderStart, postorderStart, leftSize);
        TreeNode rightChild = buildTree(inorder, postorder, inorderIndex + 1, postorderStart + leftSize, rightSize);
        
        newRoot.left = leftChild;
        newRoot.right = rightChild;
        
        return newRoot;
    }
}

File: Leetcode 0106 Construct Binary Tree from Inorder and Postorder Traversal [medium] [tree]/Construct Binary Tree from Inorder and Postorder Traversal [divide-conquer].js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
var buildTree = function(inorder, postorder) {
    if (inorder === null || postorder === null || inorder.length !== postorder.length) {
        return null;
    }
    return createTree(inorder, postorder, 0, 0, inorder.length);
};

let createTree = (inorder, postorder, inorderStart, postorderStart, size) => {
    if (size === 0) {
        return null;
    }
    
    let newRoot = new TreeNode(postorder[postorderStart + size - 1]);
    let inorderIndex = inorderStart;
    while (inorder[inorderIndex] !== postorder[postorderStart + size - 1]) {
        inorderIndex++;
    }
    
    let leftSize = inorderIndex - inorderStart;
    let rightSize = size - leftSize - 1;
    
    let leftChild = createTree(inorder, postorder, inorderStart, postorderStart, leftSize);
    let rightChild = createTree(inorder, postorder, inorderIndex + 1, postorderStart + leftSize, rightSize);
    
    newRoot.left = leftChild;
    newRoot.right = rightChild;
    return newRoot;
}
